### PR TITLE
KEYCLOAK-6473 KEYCLOAK-6472 SAML parser refactor + protocol parsers

### PR DESCRIPTION
**Pull request by hmlnarik**

### CODE REVIEW

File: saml-core/src/main/java/org/keycloak/saml/common/parsers/AbstractStaxParser.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    protected boolean isUnknownElement(E token) {
+        return token == null || Objects.equals(token, unknownElement);
+    }
+
+    protected abstract E getElementFromName(QName name);
+
+    /**
+     * Instantiates the target Java class representing the current element.<br>
+     * <b>Precondition:</b> Current event is the {@link StartElement}<br>
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/common/parsers/AbstractStaxParser.java]: Need to check - to be consistent with other standard StAX javadocs.

File: saml-core/src/main/java/org/keycloak/saml/common/util/StaxParserUtil.java [hunk 2]
```diff
@@ -263,7 +363,9 @@ public static Element getDOMElement(XMLEventReader xmlEventReader) throws Parsin
     }
 
     /**
-     * Get the element text.
+     * Get the element text. Following {@link XMLEventReader#getElementText()}:
+     * Precondition: the current event is START_ELEMENT.
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/common/util/StaxParserUtil.java]: Need to check - to be consistent with other standard StAX javadocs.

File: saml-core/src/main/java/org/keycloak/saml/common/util/StaxParserUtil.java [hunk 3]
```diff
@@ -411,7 +513,7 @@ public static EndElement getNextEndElement(XMLEventReader xmlEventReader) throws
      *
      * @return
      */
-    public static String getStartElementName(StartElement startElement) {
+    public static String getElementName(StartElement startElement) {
```
Comment (pskopek) [inline saml-core/src/main/java/org/keycloak/saml/common/util/StaxParserUtil.java]: This is just slight API change which is not customer facing, so we can leave it.

File: saml-core/src/main/java/org/keycloak/saml/common/util/StaxParserUtil.java [hunk 4]
```diff
@@ -422,10 +524,12 @@ public static String getStartElementName(StartElement startElement) {
      *
      * @return
      */
-    public static String getEndElementName(EndElement endElement) {
+    public static String getElementName(EndElement endElement) {
```
Comment (pskopek) [inline saml-core/src/main/java/org/keycloak/saml/common/util/StaxParserUtil.java]: The same as getStartElementName, we can leave this API change as suggested.

**Review by pskopek — APPROVED** [review]
Just minor API changes which are not customer facing.

File: saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/assertion/SAMLAttributeValueParser.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.saml.processing.core.parsers.saml.assertion;
+
+import org.keycloak.saml.common.PicketLinkLogger;
+import org.keycloak.saml.common.PicketLinkLoggerFactory;
+import org.keycloak.saml.common.constants.JBossSAMLConstants;
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/assertion/SAMLAttributeValueParser.java]: Using {{JBossSAMLConstants}} should be avoided

**Review by pskopek — APPROVED** [review]
All new changes are OK. Approved.

**Review by mposolda — APPROVED** [review]

**Pull request merged**